src/MediatR/Mediator.cs

```diff
@@ -48,21 +42,8 @@ public Task Publish<TNotification>(TNotification notification, CancellationToken
                 throw new ArgumentNullException(nameof(notification));
             }
 
-            var notificationType = notification.GetType();
-            var handler = _notificationHandlers.GetOrAdd(notificationType,
-                t => (NotificationHandlerWrapper)Activator.CreateInstance(typeof(NotificationHandlerWrapperImpl<>).MakeGenericType(notificationType)));
-
-            return handler.Handle(notification, cancellationToken, _serviceFactory, PublishCore);
-        }
-
-        /// <summary>
-        /// Override in a derived class to control how the tasks are awaited. By default the implementation is <see cref="Task.WhenAll(IEnumerable{Task})" />
-        /// </summary>
-        /// <param name="allHandlers">Enumerable of tasks representing invoking each notification handler</param>
-        /// <returns>A task representing invoking all handlers</returns>
-        protected virtual Task PublishCore(IEnumerable<Task> allHandlers)
-        {
-            return Task.WhenAll(allHandlers);
+            var mediator = (INotificationMediator<TNotification>)_serviceFactory(typeof(INotificationMediator<TNotification>));

```

Comment:
This should be 'notification.GetType() / MakeGenericType' like the original code and how Send does it. So you don't have to pass concrete types (e.g. if you pass an 'IEvent' it resolves the actual type of the object, not of IEvent).